pkg/ccl/changefeedccl/client_bench_test.go

```diff
@@ -0,0 +1,592 @@
+	writer.WriteHeader(200)
+}
+
+func (cb *changefeedBenchmark) setupMux() {
+	m := http.NewServeMux()
+	m.HandleFunc(serverStartPath, cb.handleServerStart)
+	m.HandleFunc(sinkStartPath, cb.handleSinkStart)
+	m.HandleFunc(sinkDonePath, cb.handleSinkDone)
+	cb.httpServer.Handler = m
+}
+
+// swallowError is a helper to swallow errors from functions used with an
+// errgroup to ensure that an error is not returned if the context has been
+// canceled.
+func swallowError(ctx context.Context, f func() error) func() error {
+	return func() error {
+		if err := f(); err != nil && ctx.Err() == nil {
+			return err
+		}
+		return nil
+	}
+}
+
+// cleanup should be called at the end of a test.
+func (cb *changefeedBenchmark) cleanup() {
+	if cb.localServer != nil {
+		cb.localServer.Stopper().Stop(context.Background())
+	}
+	if cb.cleanupFunc != nil {
+		cb.cleanupFunc()
+	}
+	os.RemoveAll(cb.tempDir)
+}
+
+func (cb *changefeedBenchmark) runCockroachServer(ctx context.Context) error {
+	serverProc := helperProcess(ctx, "server",
+		"-n", strconv.Itoa(cb.n),
+		"--test-addr", cb.httpListener.Addr().String())
+	cb.serverSubprocess.done = make(chan struct{})
+	defer close(cb.serverSubprocess.done)
+	if err := serverProc.Run(); err != nil && ctx.Err() == nil {
+		return errors.Wrapf(err, "output: %s", "out")
+	}
+	return nil
+}
+
+func (cb *changefeedBenchmark) runSink(ctx context.Context) error {
+	pgURL, cleanup := sqlutils.PGUrl(cb.tb, cb.localServer.ServingSQLAddr(), "", url.User(security.RootUser))
+	cb.cleanupFunc = cleanup
+
+	// Let's now set up some data
+	var sink *exec.Cmd
+	args := []string{
+		"-n", strconv.Itoa(cb.n),
+		"--pgurl", pgURL.String(),
+		"--test-addr", cb.httpListener.Addr().String(),
+	}
+	switch cb.sinkType {
+	case sinklessSink:
+		sink = helperProcess(ctx, "sink-sinkless", args...)
+	case httpSink:
+		sink = helperProcess(ctx, "sink-http", args...)
+	case kafkaSink:
+		return fmt.Errorf("kafka sink not yet support")
+	default:
+		return fmt.Errorf("unknown sink type %v", cb.sinkType)
+	}
+
+	if err := sink.Run(); err != nil {
+		return errors.Wrapf(err, "output: %s", "out")
+	}
+	return nil
+}
+
+func (cb *changefeedBenchmark) runLocalServer() {
+	spec := base.DefaultTestStoreSpec
+	spec.Attributes.Attrs = append(spec.Attributes.Attrs, "coordinator")
+	tc := testcluster.StartTestCluster(cb.tb, 1, base.TestClusterArgs{})
+	socketFile := filepath.Join(cb.tempDir, "test.crdb.socket")
+	tc.StopServer(0) // This is a horrible hack to get a testServer
+	tc.AddServer(cb.tb, base.TestServerArgs{
+		Addr:          "127.0.0.1:0",
+		PartOfCluster: true,
+		SSLCertsDir:   cb.serverSubprocess.sslCertsDir,
+		JoinAddr:      cb.serverSubprocess.rpcAddr,
+		StoreSpecs:    []base.StoreSpec{spec},
+		SocketFile:    socketFile,
+	})
+	cb.localServer = tc.Server(1)
+}
+
+func helperProcess(ctx context.Context, c string, args ...string) *exec.Cmd {
+	cs := append(os.Args[1:], append([]string{"-test.v", "-test.run=TestHelper", "--", c}, args...)...)
```

Comment:
this feels pretty hacky. It depends on us not running our tests with `ENVVAR=... go test` right?